src/stan/variational/advi.hpp

```diff
@@ -0,0 +1,665 @@
+                      << std::right << std::setw(9) << std::setprecision(1)
+                      << elbo
+                      << "  "
+                      << std::setw(16) << std::fixed << std::setprecision(3)
+                      << delta_elbo_ave
+                      << "  "
+                      << std::setw(15) << std::fixed << std::setprecision(3)
+                      << delta_elbo_med;
+
+            if (diag_stream_) {
+              end = clock();
+              delta_t = static_cast<double>(end - start) / CLOCKS_PER_SEC;
+
+              print_vector.clear();
+              print_vector.push_back(delta_t);
+              print_vector.push_back(elbo);
+              services::io::write_iteration_csv(*diag_stream_,
+                                                iter_counter, print_vector);
+            }
+
+            if (delta_elbo_ave < tol_rel_obj) {
+              *print_stream_ << "   MEAN ELBO CONVERGED";
+              do_more_iterations = false;
+            }
+
+            if (delta_elbo_med < tol_rel_obj) {
+              *print_stream_ << "   MEDIAN ELBO CONVERGED";
+              do_more_iterations = false;
+            }
+
+            if (iter_counter > 100) {
+              if (delta_elbo_med > 0.5 || delta_elbo_ave > 0.5) {
+                *print_stream_ << "   MAY BE DIVERGING... INSPECT ELBO";
+              }
+            }
+
+            *print_stream_ << std::endl;
+          }
+
+          // Check for max iterations
+          if (iter_counter == max_iterations) {
+            *print_stream_ << "MAX ITERATIONS REACHED" << std::endl;
+            do_more_iterations = false;
+          }
+
+          ++iter_counter;
+        }
+      }
+
+      void run_fullrank(double tol_rel_obj, int max_iterations) {
+        *print_stream_
+        << "This is stan::variational::advi::run_fullrank()" << std::endl;
+
+        // Initialize variational parameters: mu, L
+        Eigen::VectorXd mu = cont_params_;
+        Eigen::MatrixXd L  = Eigen::MatrixXd::Identity(model_.num_params_r(),
+                                                       model_.num_params_r());
+        advi_params_fullrank muL = advi_params_fullrank(mu, L);
+
+        // Robbins Monro ADAgrad
+        do_robbins_monro_adagrad(muL, tol_rel_obj, max_iterations);
+
+        cont_params_ = muL.mu();
+
+        // *print_stream_
+        // << "mu = " << std::endl
+        // << muL.mu() << std::endl;
+
+        // *print_stream_
+        // << "Sigma = " << std::endl
+        // << muL.L_chol() * muL.L_chol().transpose() << std::endl;
+
+        // std::stringstream s;
+        // stan::io::dump_writer writer(s);
+        // writer.write("mu", muL.mu());
+        // s << "\n";
+        // writer.write("L_chol", muL.L_chol());
+        // std::string written = s.str();
+        // *print_stream_ << written << std::endl;
+
+        return;
+      }
+
+      void run_meanfield(double tol_rel_obj, int max_iterations) {
```

Comment:
@akucukelbir Do we need these functions? It doesn't look like this or run_fullrank do very much aside from calling two child functions.